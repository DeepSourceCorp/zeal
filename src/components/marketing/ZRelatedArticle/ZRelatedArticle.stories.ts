import ZRelatedArticle from './ZRelatedArticle.vue'

export default {
  title: 'Zeal Marketing/Related Article',
  component: ZRelatedArticle,
  excludeStories: /.*Data$/
}

export const Basic = () => ({
  components: { ZRelatedArticle },
  template: `<div class='padded-container bg-ink-400 text-vanilla-400 grid grid-cols-12'>
      <z-related-article class="col-span-8">
        <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
        <a href="/somethign" slot="link">Read More</a>
      </z-related-article>
    </div>`
})