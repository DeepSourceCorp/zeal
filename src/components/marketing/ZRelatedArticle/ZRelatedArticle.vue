<template>
  <span class="grid grid-cols-12 py-3">
    <span class="col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9">
      <slot name="heading"></slot>
    </span>
    <span class="flex items-center col-span-10 text-sm font-medium uppercase text-juniper">
      <slot name="link"></slot>
      <z-icon color="juniper" class="ml-1" size="small" icon="arrow-right"></z-icon>
    </span>
  </span>
</template>

<script lang="ts">
import Vue from 'vue'
import ZIcon from '../../ZIcon'

export default Vue.extend({
  name: 'ZRelatedArticle',
  components: {
    ZIcon
  }
})
</script>
