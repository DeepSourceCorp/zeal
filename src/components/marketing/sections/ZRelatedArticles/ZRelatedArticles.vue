<template>
  <div>
    <span class="block mb-3 text-sm font-medium uppercase text-vanilla-400">{{ header }}</span>
    <div class="border-t border-b border-slate md:grid grid-cols-12">
      <slot></slot>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

const SECONDARY_ARTICLE_CLASSES = ['md:pl-4', 'md:border-l', 'md:border-t-0', 'border-t']
const COMMON_ARTICLE_CLASSES = ['col-span-6']

export default Vue.extend({
  name: 'ZRelatedArticles',
  props: {
    header: String
  },
  data() {
    return {
      articles: [] as Vue[]
    }
  },
  mounted() {
    this.acquireFirstTwoArticlesFromChildren()
    this.assignStylesToSecondaryArticles()
    this.assignCommonStylesToAllArticles()
  },
  methods: {
    acquireFirstTwoArticlesFromChildren() {
      /**
       * Acquires all the child components that are Related Article components,
       * and stores the first two of them.
       */
      this.articles = this.$children.filter(({ $options }) => $options.name === 'ZRelatedArticle').slice(0, 2)
    },
    assignStylesToSecondaryArticles() {
      /**
       * Assigns custom secondary tyling classes, to the secondary articles
       * (articles except the first one).
       */
      this.articles.map(({ $el }, index) => {
        if (index > 0) {
          $el.classList.add(...SECONDARY_ARTICLE_CLASSES)
        }
      })
    },
    assignCommonStylesToAllArticles() {
      /**
       * Assigns custom common styling classes to all the articles.
       */
      this.articles.map(({ $el }) => {
        $el.classList.add(...COMMON_ARTICLE_CLASSES)
      })
    }
  }
})
</script>
