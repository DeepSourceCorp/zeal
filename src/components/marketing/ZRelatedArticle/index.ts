import ZRelatedArticle from './ZRelatedArticle.vue'

export default ZRelatedArticle
