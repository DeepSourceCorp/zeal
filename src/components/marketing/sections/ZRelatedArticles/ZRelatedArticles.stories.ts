import ZRelatedArticles from './ZRelatedArticles.vue'
import ZRelatedArticle from '../../ZRelatedArticle'

export default {
  title: 'Zeal Marketing/Related Articles',
  component: ZRelatedArticles,
  excludeStories: /.*Data$/
}

export const WithBothArticles = () => ({
  components: { ZRelatedArticles, ZRelatedArticle },
  template: `<div class='padded-container bg-ink-400 text-vanilla-400 grid grid-cols-12'>
      <z-related-articles header="Keep reading" class="col-span-8">
        <z-related-article>
          <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
          <a href="/somethign" slot="link">Read More</a>
        </z-related-article>
        <z-related-article>
          <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
          <a href="/somethign" slot="link">Read More</a>
        </z-related-article>
      </z-related-articles>
    </div>`
})

export const WithOnlyFirstArticle = () => ({
  components: { ZRelatedArticles, ZRelatedArticle },
  template: `<div class='padded-container bg-ink-400 text-vanilla-400 grid grid-cols-12'>
      <z-related-articles header="Keep reading" class="col-span-8">
        <z-related-article>
          <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
          <a href="/somethign" slot="link">Read More</a>
        </z-related-article>
      </z-related-articles>
    </div>`
})
