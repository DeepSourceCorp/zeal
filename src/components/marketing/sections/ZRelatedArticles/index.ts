import ZRelatedArticles from './ZRelatedArticles.vue'

export default ZRelatedArticles
