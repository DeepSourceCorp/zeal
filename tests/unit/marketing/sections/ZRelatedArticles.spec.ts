import Vue from 'vue'
import { createLocalVue, mount, Wrapper } from '@vue/test-utils'

import ZRelatedArticles from '../../../../src/components/marketing/sections/ZRelatedArticles'
import ZRelatedArticle from '../../../../src/components/marketing/ZRelatedArticle'

describe('Page Label component', () => {
  let mountFunction: (options?: Record<string, unknown>) => Wrapper<Vue>
  let localVue: typeof Vue
  beforeEach(() => {
    localVue = createLocalVue()
    mountFunction = (options = {}) => {
      return mount(ZRelatedArticles, {
        localVue,
        ...options
      })
    }
  })
  it('renders the section once its mounted', () => {
    expect(mountFunction().html()).toMatchSnapshot()
  })

  it('renders only two related articles even on several input', () => {
    let mountedInstance = mountFunction({
      components: {
        ZRelatedArticle
      },
      slots: {
        default: [
          `<z-related-article>
            <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
            <a href="/somethign" slot="link">Read More</a>
          </z-related-article>`,
          `<z-related-article>
            <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
            <a href="/somethign" slot="link">Read More</a>
          </z-related-article>`,
          `<z-related-article>
            <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
            <a href="/somethign" slot="link">Read More</a>
          </z-related-article>`
        ]
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.vm.$data.articles.length).toEqual(2)
  })

  it('secondary articles have the correct styling classes', () => {
    let mountedInstance = mountFunction({
      components: {
        ZRelatedArticle
      },
      slots: {
        default: [
          `<z-related-article>
            <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
            <a href="/somethign" slot="link">Read More</a>
          </z-related-article>`,
          `<z-related-article>
            <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
            <a href="/somethign" slot="link">Read More</a>
          </z-related-article>`
        ]
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.vm.$data.articles[1].$el.className.includes('pl-4 border-l')).toBe(true)
  })

  it('render no vertical border when only one article is available', () => {
    let mountedInstance = mountFunction({
      components: {
        ZRelatedArticle
      },
      slots: {
        default: [
          `<z-related-article>
            <span slot="heading">Odio amet tincidunt metus dictum dapibus cum nibh sed</span>
            <a href="/somethign" slot="link">Read More</a>
          </z-related-article>`
        ]
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.vm.$data.articles[0].$el.className.includes('pl-4 border-l')).toBe(false)
  })
})
