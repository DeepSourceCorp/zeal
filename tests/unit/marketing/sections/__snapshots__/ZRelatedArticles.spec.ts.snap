// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Articles section render no vertical border when only one article is available 1`] = `
<div><span class="block mb-3 text-sm font-medium uppercase text-vanilla-400"></span>
  <div class="border-t border-b border-slate md:grid grid-cols-12"><span class="block sm:grid grid-cols-12 py-3 col-span-6"><span class="sm:col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9"><span>Odio amet tincidunt metus dictum dapibus cum nibh sed</span></span> <span class="flex items-center sm:col-span-10 text-sm font-medium uppercase text-juniper"><a href="/somethign">Read More</a> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="ml-1 z-icon stroke-1.5 w-4 h-4 z-icon--arrow-right text-juniper"><line x1="5" y1="12" x2="19" y2="12"></line><polyline points="12 5 19 12 12 19"></polyline></svg></span></span></div>
</div>
`;

exports[`Related Articles section renders only two related articles even on several input 1`] = `
<div><span class="block mb-3 text-sm font-medium uppercase text-vanilla-400"></span>
  <div class="border-t border-b border-slate md:grid grid-cols-12"><span class="block sm:grid grid-cols-12 py-3 col-span-6"><span class="sm:col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9"><span>Odio amet tincidunt metus dictum dapibus cum nibh sed</span></span> <span class="flex items-center sm:col-span-10 text-sm font-medium uppercase text-juniper"><a href="/somethign">Read More</a> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="ml-1 z-icon stroke-1.5 w-4 h-4 z-icon--arrow-right text-juniper"><line x1="5" y1="12" x2="19" y2="12"></line><polyline points="12 5 19 12 12 19"></polyline></svg></span></span><span class="block sm:grid grid-cols-12 py-3 md:pl-4 md:border-l md:border-t-0 border-t col-span-6"><span class="sm:col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9"><span>Odio amet tincidunt metus dictum dapibus cum nibh sed</span></span> <span class="flex items-center sm:col-span-10 text-sm font-medium uppercase text-juniper"><a href="/somethign">Read More</a> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="ml-1 z-icon stroke-1.5 w-4 h-4 z-icon--arrow-right text-juniper"><line x1="5" y1="12" x2="19" y2="12"></line><polyline points="12 5 19 12 12 19"></polyline></svg></span></span><span class="block sm:grid grid-cols-12 py-3"><span class="sm:col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9"><span>Odio amet tincidunt metus dictum dapibus cum nibh sed</span></span> <span class="flex items-center sm:col-span-10 text-sm font-medium uppercase text-juniper"><a href="/somethign">Read More</a> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="ml-1 z-icon stroke-1.5 w-4 h-4 z-icon--arrow-right text-juniper"><line x1="5" y1="12" x2="19" y2="12"></line><polyline points="12 5 19 12 12 19"></polyline></svg></span></span></div>
</div>
`;

exports[`Related Articles section renders the section once its mounted 1`] = `
<div><span class="block mb-3 text-sm font-medium uppercase text-vanilla-400"></span>
  <div class="border-t border-b border-slate md:grid grid-cols-12"></div>
</div>
`;

exports[`Related Articles section secondary articles have the correct styling classes 1`] = `
<div><span class="block mb-3 text-sm font-medium uppercase text-vanilla-400"></span>
  <div class="border-t border-b border-slate md:grid grid-cols-12"><span class="block sm:grid grid-cols-12 py-3 col-span-6"><span class="sm:col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9"><span>Odio amet tincidunt metus dictum dapibus cum nibh sed</span></span> <span class="flex items-center sm:col-span-10 text-sm font-medium uppercase text-juniper"><a href="/somethign">Read More</a> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="ml-1 z-icon stroke-1.5 w-4 h-4 z-icon--arrow-right text-juniper"><line x1="5" y1="12" x2="19" y2="12"></line><polyline points="12 5 19 12 12 19"></polyline></svg></span></span><span class="block sm:grid grid-cols-12 py-3 md:pl-4 md:border-l md:border-t-0 border-t col-span-6"><span class="sm:col-span-10 mb-2 block text-vanilla-200 text-lg font-medium leading-9"><span>Odio amet tincidunt metus dictum dapibus cum nibh sed</span></span> <span class="flex items-center sm:col-span-10 text-sm font-medium uppercase text-juniper"><a href="/somethign">Read More</a> <svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="ml-1 z-icon stroke-1.5 w-4 h-4 z-icon--arrow-right text-juniper"><line x1="5" y1="12" x2="19" y2="12"></line><polyline points="12 5 19 12 12 19"></polyline></svg></span></span></div>
</div>
`;
